lib/widget/toggle_button.dart

```diff
@@ -0,0 +1,303 @@
+      'color': (value) => _controller.color = Utils.getColor(value),
+      'selectedColor': (value) =>
+          _controller.selectedColor = Utils.getColor(value),
+      'selectedBackgroundColor': (value) =>
+          _controller.selectedBackgroundColor = Utils.getColor(value),
+      'borderColor': (value) => _controller.borderColor = Utils.getColor(value),
+      'selectedBorderColor': (value) =>
+          _controller.selectedBorderColor = Utils.getColor(value),
+      'onChange': (definition) => _controller.onChange =
+          framework.EnsembleAction.fromYaml(definition, initiator: this),
+    };
+  }
+
+  @override
+  Map<String, Function> methods() {
+    return {};
+  }
+
+  bool isValueInItems() {
+    if (_controller.maybeValue != null && _controller.items != null) {
+      for (SelectOneItem item in _controller.items!) {
+        if (_controller.maybeValue == item.value) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  dynamic getValue() {
+    if (isValueInItems()) {
+      return _controller.maybeValue;
+    }
+    return null;
+  }
+
+  void updateItems(dynamic values) {
+    List<SelectOneItem> entries = [];
+    if (values is List) {
+      calculateIconSize(values);
+      for (var element in values) {
+        // must be of value/label pair. Maybe let user overrides later
+        if (element is Map) {
+          if (element['value'] != null) {
+            entries.add(
+              SelectOneItem(
+                value: element['value'],
+                label: element['label']?.toString(),
+                icon: Utils.getIcon(element['icon']),
+                isIcon: isIconExist(values),
+              ),
+            );
+          }
+        }
+        // simply use the value
+        else {
+          entries.add(SelectOneItem(value: element));
+        }
+      }
+    }
+    _controller.items = entries;
+
+    // ensure that the value is still correct
+    if (!isValueInItems()) {
+      _controller.maybeValue = null;
+    }
+  }
+
+  bool isIconExist(List v) {
+    for (var e in v) {
+      if (e['icon'] != null) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void calculateIconSize(List v) {
+    List iconSize = [];
+    _controller.gap = 0;
+    for (var e in v) {
+      if (e is Map) {
+        if (e['icon'] != null) {
+          if (e['icon']['size'] != null) {
+            iconSize.add(e['icon']['size']);
+          }
+        }
+      }
+    }
+    if (iconSize.isNotEmpty) {
+      _controller.gap =
+          iconSize.reduce((curr, next) => curr > next ? curr : next);
+    } else {
+      _controller.gap = 0;
+    }
+  }
+
+  void onSelectionChanged(dynamic value) {
+    setProperty('value', value);
+  }
+}
+
+class EnsembleToggleButtonState extends WidgetState<EnsembleToggleButton> {
+  List<ToggleItem>? _items = [];
+
+  @override
+  void initState() {
+    super.initState();
+    _setValue();
+  }
+
+  void _setValue() {
```

Comment:
Why do we need both `_setValue` and `_updatedSelectedState`?

Reply:
`_setValue` is for creating an initial items list and its selection based on the value. `_updatedSelectedState` is used to update the existing list and its selection state based on the index value